reconcile/change_owners/change_owners.py

```diff
@@ -231,6 +242,31 @@ def init_gitlab(gitlab_project_id: str) -> GitLabApi:
     return GitLabApi(instance, project_id=gitlab_project_id, settings=settings)
 
 
+def is_change_admitted(
+    changes: list[BundleFileChange], user: str, good_to_test_approvers: set[str]
+) -> bool:
+    for change in changes:
+        for dc in change.diff_coverage:
+            for c in dc.coverage:
+                if c.change_type_processor.restrictive:
+                    approvers = {a.org_username for a in c.approvers}
+
+                    for gttapprover in good_to_test_approvers:
+                        if gttapprover in approvers:
+                            logging.info(
+                                f"Restrictive change approved by {gttapprover}"
+                            )
+                            return True
+
+                    if user not in approvers:
+                        logging.error(
+                            f"change type {c.change_type_processor.name} is restrictive"
+                            f"user {user} is not an approver"
+                        )
+                        return False
```

Comment:
`user` here is the author, why check whether a change is admitted requires checking author? I thought the logic is just all restrictive changes have good to test approved. Like this:

```python
    return all(
        {a.org_username for a in c.approvers} & good_to_test_approvers
        for change in changes
        for dc in change.diff_coverage
        for c in dc.coverage
        if c.change_type_processor.restrictive
    )

```

Reply:
No it's more complex: Users with the restrictive role assigned are allowed to do these restrictive changes and add the looks good to test comment.

Reply:
A change is either admitted by schema or by a comment of an authorized person. I think the name is fine.

Reply:
What do you mean by "admitted by schema"? I don't really see where in this function only the schema check takes place (most likely is my newbiness to this integration).

I share a bit the feeling of not understanding well this function. The relationship between the `user`, the `approvers` and the `good_to_test_approvers` is complex.

Reply:
I may understand what you mean. Unlike `lgtm` flow, even the author is part of approvers, still need a `lgtm` to mark it as approved. But for this admission check, if author is part of approvers, then it's admitted implicitly, right?

Btw this code is not covering all cases due to use of early `return`.

If the change has 2 different coverages, each one has a different list of approvers, one is `[A]`, the other one is `[B]`, only `A` commented good to test, then this change should be considered not admitted, require `B` to comment good to test.

To describe the condition:

* changes are admitted when all restrictive coverages are admitted
* a coverage is admitted when the author is part of approvers OR some approvers commented good to test

Translate to code:

```python

def is_coverage_admitted(coverage, user, good_to_test_approvers):
    return any(
        a.org_username == user or a.org_username in good_to_test_approvers
        for a in coverage.approvers
    )


def is_change_admitted(
    changes: list[BundleFileChange], user: str, good_to_test_approvers: set[str]
) -> bool:
    return all(
        is_coverage_admitted(c, user, good_to_test_approvers)
        for change in changes
        for dc in change.diff_coverage
        for c in dc.coverage
        if c.change_type_processor.restrictive
    )

```

Maybe should rename `user` to `author` to make it explicit.

Reply:
updated